Source/WebCore/html/InputType.cpp

```diff
@@ -169,6 +169,81 @@ bool InputType::themeSupportsDataListUI(InputType* type)
     return RenderTheme::singleton().supportsDataListUI(type->formControlType());
 }
 
+template<typename T> static bool validateInputType(const T& inputType, const String& value)
+{
+    ASSERT(inputType.canSetStringValue());
+    return !inputType.typeMismatchFor(value)
+        && !inputType.stepMismatch(value)
+        && !inputType.rangeUnderflow(value)
+        && !inputType.rangeOverflow(value)
+        && !inputType.patternMismatch(value)
+        && !inputType.valueMissing(value);
+}
+
+bool InputType::isValidValue(const String& value) const
+{
+    switch (m_type) {
+    case Type::Button:
+        return validateInputType(downcast<ButtonInputType>(*this), value);
+    case Type::Checkbox:
+        return validateInputType(downcast<CheckboxInputType>(*this), value);
```

Comment:
Was wondering to myself whether

```c++
if (is<ButtonInputType>(*this)))
    return validateInputType(downcast<ButtonInputType>(*this), value);

if (is<CheckboxInputType>(*this))
    return validateInputType(downcast<CheckboxInputType(*this), value);

...
```

might read better (assuming the compiler generates the same code, which I expect it would), but upon reflection think the switch statement wins out.

Reply:
Depending on how clever you want to be, you could write a templated function that uses a parameter pack to iterate over the concrete input types, like (untested)

```c++
template <typename... Type>
void doIt(const InputType& inputType, const String& value)
{
    (is<Type>(inputType) && (validateInputType(downcast<Type>(inputType), value), true) || ...;
}

bool InputType::isValidValue(const String& value) const
{
    doIt<
        ButtonInputType,
        CheckboxInputType,
#if ENABLE(INPUT_TYPE_COLOR)
        ColorInputType,
#endif
        ...
    >(*this, value);
}
```

Maybe there's a better way to write that. (Maybe there's already something in WTF to let you loop over a list of types?)

Reply:
Are you sure both versions generate the same code? We want a jump table here.

Reply:
I did say "assuming the compiler generates the same code". I have seen previously the compiler generate jump tables for sequences of if statements that check a value like this, but of course it would be have to be checked.

Reply:
My version was carefully A/B tested twice. If I change it, I'd have to redo it. I am not sure the proposed code is that much nicer that it seems worth it TBH.

Reply:
Also note that another good reason for using a switch is that you get a build error if you introduce a new input type and fail to update it.

Reply:
True, the build error is a good reason to stick with the switch statement.